app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/widget/MVVWidget.kt

```diff
@@ -0,0 +1,184 @@
+package de.tum.`in`.tumcampusapp.component.ui.transportation.widget
+
+import android.app.AlarmManager
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.view.View
+import android.widget.RemoteViews
+import de.tum.`in`.tumcampusapp.R
+import de.tum.`in`.tumcampusapp.component.ui.transportation.TransportController
+import java.util.*
+
+/**
+ * Implementation of App Widget functionality.
+ * App Widget Configuration implemented in [MVVWidgetConfigureActivity]
+ */
+class MVVWidget : AppWidgetProvider() {
+
+    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
+        updateAppWidgets(context, appWidgetManager, appWidgetIds)
+        setAlarm(context)
+        super.onUpdate(context, appWidgetManager, appWidgetIds)
+    }
+
+    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
+        // When the user deletes the widget, delete the associated setting from the database.
+        for (appWidgetId in appWidgetIds) {
+            TransportController(context).deleteWidget(appWidgetId)
+        }
+        super.onDeleted(context, appWidgetIds)
+    }
+
+    override fun onDisabled(context: Context) {
+        // Cancel alarm as the last widget has been removed
+        setAlarm(context)
+        super.onDisabled(context)
+    }
+
+    /**
+     * If no alarm is running yet a new alarm is started which repeats every minute
+     */
+    private fun setAlarm(context: Context) {
+        var autoReload = false
+
+        for (appWidgetId in getActiveWidgetIds(context)) {
+            val widgetDepartures = TransportController(context).getWidget(appWidgetId)
+            if (widgetDepartures.autoReload) {
+                autoReload = true
+                break
+            }
+        }
+
+        val intent = Intent(context, MVVWidget::class.java)
+        val sender = PendingIntent.getBroadcast(context, 0, intent, 0)
+        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager?
+        alarmManager?.let {
+            it.cancel(sender)
+            if (autoReload) {
+                intent.action = BROADCAST_RELOAD_ALL_ALARM
+                it.setRepeating(AlarmManager.RTC, 5000, UPDATE_ALARM_DELAY.toLong(), sender)
+            }
+        }
+    }
+
+    /**
+     * Plans updates the widgets after 30s and 60s
```

Comment:
The comment seems to be out of sync with the implementation, which refreshes every 20 seconds.